File: 2. Medium/173. Binary Search Tree Iterator/Solution.java
class BSTIterator {
    private TreeNode root;
    private Stack<TreeNode> stack;

    public BSTIterator(TreeNode root) {
        this.root = root;
        stack = new Stack<>();
    }

    public int next() {
        while (root != null) {
            stack.push(root);
            root = root.left;
        }
        root = stack.pop();
        int res = root.val;
        root = root.right;
        return res;
    }

    public boolean hasNext() {
        return root != null || !stack.isEmpty();
    }
}

File: 2. Medium/173. Binary Search Tree Iterator/solution.js
/**
 * @param {TreeNode} root
 */
var BSTIterator = function (root) {
    this.stack = [];
    this.root = root;
};

/**
 * @return {number}
 */
BSTIterator.prototype.next = function () {
    while (this.root) {
        this.stack.push(this.root);
        this.root = this.root.left;
    }
    this.root = this.stack.pop();
    let res = this.root.val;
    this.root = this.root.right;
    return res;
};

/**
 * @return {boolean}
 */
BSTIterator.prototype.hasNext = function () {
    return this.root || this.stack.length > 0;
};